server/sendEthernet.go

```diff
@@ -0,0 +1,93 @@
+package server
+
+//function from https://gist.github.com/corny/5e4e3f8e6f2395726e46c3db9db17f12#file-dhcp_discover-go
+import (
+	"encoding/binary"
+	"net"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/google/gopacket/pcap"
+	"github.com/insomniacslk/dhcp/dhcpv4"
+)
+
+//this function sends an unicast to the hardware address defined in resp.ClientHWAddr,
+//the layer3 destination address is still the broadcast address;
+//iface: the interface where the DHCP message should be sent;
+//resp: DHCPv4 struct, which should be sent;
+func sendEthernet(iface net.Interface, resp *dhcpv4.DHCPv4) {
+
+	eth := layers.Ethernet{
+		EthernetType: layers.EthernetTypeIPv4,
+		SrcMAC:       iface.HardwareAddr,
+		DstMAC:       resp.ClientHWAddr,
+	}
+	ip := layers.IPv4{
+		Version:  4,
+		TTL:      64,
+		SrcIP:    resp.ServerIPAddr,
+		DstIP:    net.IPv4bcast,
+		Protocol: layers.IPProtocolUDP,
+	}
+	udp := layers.UDP{
+		SrcPort: dhcpv4.ServerPort,
+		DstPort: dhcpv4.ClientPort,
+	}
+
+	//put all data from request in the layer struct
+	dhcp := layers.DHCPv4{
+		Operation:    layers.DHCPOp(resp.OpCode),
+		HardwareType: layers.LinkType(resp.HWType),
+		HardwareOpts: resp.HopCount,
+		Xid:          binary.BigEndian.Uint32(resp.TransactionID[:]),
+		Secs:         resp.NumSeconds,
+		Flags:        resp.Flags,
+		ClientIP:     resp.ClientIPAddr,
+		YourClientIP: resp.YourIPAddr,
+		NextServerIP: resp.ServerIPAddr,
+		RelayAgentIP: resp.GatewayIPAddr,
+		ClientHWAddr: resp.ClientHWAddr,
+		ServerName:   []byte(resp.ServerHostName),
+	}
+
+	appendOption := func(optType layers.DHCPOpt, data []byte) {
+		dhcp.Options = append(dhcp.Options, layers.DHCPOption{
+			Type:   optType,
+			Data:   data,
+			Length: uint8(len(data)),
+		})
+		return
+	}
+
+	//Add all option to the layer struct
+	for key, element := range resp.Options {
+		appendOption(layers.DHCPOpt(key), []byte(element))
+	}
```

Comment:
Since the dhcp library we use already has serialization code for dhcp, can we replace this with just the payload as a byte slice (from `resp.ToBytes()`) ? or does this absolutely need to use the layers from gopacket for serialization ?

One reason I'd like to avoid this is that if we end up with 2 different serialization codepaths it'd make debugging painful

Reply:
Good Point! 
Now I use resp.ToBytes()